Plugins/Json/src/DetectorJsonConverter.cpp

```diff
@@ -0,0 +1,176 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2023 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "Acts/Plugins/Json/DetectorJsonConverter.hpp"
+
+#include "Acts/Detector/Detector.hpp"
+#include "Acts/Detector/DetectorVolume.hpp"
+#include "Acts/Detector/Portal.hpp"
+#include "Acts/Navigation/DetectorVolumeFinders.hpp"
+#include "Acts/Plugins/Json/DetectorVolumeJsonConverter.hpp"
+#include "Acts/Plugins/Json/DetrayJsonHelper.hpp"
+#include "Acts/Plugins/Json/PortalJsonConverter.hpp"
+#include "Acts/Utilities/Enumerate.hpp"
+#include "Acts/Utilities/Helpers.hpp"
+
+#include <ctime>
+#include <set>
+
+nlohmann::json Acts::DetectorJsonConverter::toJson(
+    const GeometryContext& gctx, const Experimental::Detector& detector,
+    const Options& options) {
+  // Get the time stamp
+  time_t tt;
+  time(&tt);
+  auto ti = localtime(&tt);
+
+  nlohmann::json jDetector;
+
+  std::size_t nSurfaces = 0;
+  std::vector<const Experimental::Portal*> portals;
+
+  for (const auto& volume : detector.volumes()) {
+    nSurfaces += volume->surfaces().size();
+    for (const auto& portal : volume->portals()) {
+      if (std::find(portals.begin(), portals.end(), portal) == portals.end()) {
+        portals.push_back(portal);
+      }
+    }
+  }
+
+  // Write the actual data objects
+  nlohmann::json jData;
+  jData["name"] = detector.name();
+
+  // The portals are written first
+  auto volumes = detector.volumes();
+  nlohmann::json jPortals;
+
+  for (const auto& portal : portals) {
+    auto jPortal = PortalJsonConverter::toJson(
+        gctx, *portal, volumes, options.volumeOptions.portalOptions);
+    jPortals.push_back(jPortal);
+  }
+  jData["portals"] = jPortals;
+
+  // The volumes are written next, with portals already defined, the volumes
+  // will only index those
+  nlohmann::json jVolumes;
+  for (const auto& volume : volumes) {
+    auto jVolume = DetectorVolumeJsonConverter::toJson(
+        gctx, *volume, volumes, portals, options.volumeOptions);
+    jVolumes.push_back(jVolume);
+  }
+  jData["volumes"] = jVolumes;
+
+  // Write the header
+  nlohmann::json jHeader;
+  jHeader["detector"] = detector.name();
+  jHeader["date"] = asctime(ti);
+  jHeader["no. volumes"] = detector.volumes().size();
```

Comment:
How about `volume_count`? Spaces in these keys are a bit awkward.

Reply:
Right, I took them from the detray header, I totally agree though.